src/Orleans.Runtime/Placement/SiloRoleBasedPlacementDirector.cs

```diff
@@ -0,0 +1,37 @@
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Orleans.Internal;
+using Orleans.Runtime.MembershipService;
+
+namespace Orleans.Runtime.Placement
+{
+    internal class SiloRoleBasedPlacementDirector : IPlacementDirector
+    {
+        private readonly MembershipTableManager membershipTableManager;
+
+        public SiloRoleBasedPlacementDirector(MembershipTableManager membershipTableManager)
+        {
+            this.membershipTableManager = membershipTableManager;
+        }
+
+        public virtual Task<SiloAddress> OnAddActivation(
+            PlacementStrategy strategy, PlacementTarget target, IPlacementContext context)
+        {
+            string siloRole = target.GrainIdentity.Key.ToString();
+
+            List<SiloAddress> siloAddressesSameRole = membershipTableManager.MembershipTableSnapshot.Entries
```

Comment:
You will need to intersect the list here form the list that you will get from `context.GetCompatibleSilos`, since some people might have heterogeneous cluster.

Reply:
done!